### PR TITLE
feat(curriculum): add registration form practice project

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Adds the _Registration Form_ practice project to the upcoming curriculum.

- [x] Merge PR fixing ERM on boundary issue

### CODE REVIEW

File: curriculum/challenges/english/01-responsive-web-design/registration-form/final.md
Comment (ShaunSHamilton) [inline curriculum/challenges/english/01-responsive-web-design/registration-form/final.md:4]: That is quite some `id` @ojeytonwilliams

Comment (nhcarrigan) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/final.md:4]: Beautiful.

Comment (ojeytonwilliams) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/final.md]: I only just noticed this, sorry.  What on Earth happened here?

Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/final.md]: @ojeytonwilliams I assumed you were to blame... I do not quite understand the significance of a `final.md` file.

**Comment by ShaunSHamilton** [comment]
I spent so long thinking this PR was already merged 😅

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-022.md
Comment (gikf) [inline curriculum/challenges/english/01-responsive-web-design/registration-form/part-022.md:10]: I've initially read it to add these texts to the labels in this part too. No issues from tests with that, but in the next part I was confused why text was gone.

Not that big of deal, as I've misread it. One way to prevent it could be in this part just mentioning two account types, without specific names. Or it could be left alone.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-039.md
Comment (gikf) [inline curriculum/challenges/english/01-responsive-web-design/registration-form/part-039.md:23]:
```suggestion
assert.equal(new __helpers.CSSHelp(document).getStyle('form')?.margin, '0px auto');
```

**Review by gikf — COMMENTED** [review]
Findings from completing project once.

More general note - there doesn't seem to be part with adding `value` attribute to the `radio` and `checkbox` inputs

Some specific things below.

**Review by moT01 — COMMENTED** [review]
Overall, this looks real good @ShaunSHamilton 🎉 

I left a number of suggestions - Many could be considered a preference and aren't necessary to add, others might just be thoughts or observations. Looks like @gikf left some suggestions for a number of the tests I was struggling with.

Some overall thoughts:
Seemed like some of the lessons could have maybe been broken up a little more. At the start I add 4 label elements at once - then 4 inputs at once, one in each - and 4 types... it's kind of a lot - I'm not saying it's wrong or needs to be changed - maybe something to keep in mind for the future. One way I might have done it is to do a single element, then the next three - for instance, 1 label in a step, then 1 input, then 1 type - followed by the last 3 labels, inputs, types... per step. Hope that makes sense.

There were some that could maybe have used more explanation - that's tough, because you don't want to get too wordy. For instance though, when a new element is introduced, say the radio buttons, you aren't really told what they do - for the most part, the elements are pretty self explanatory so it's not really that big of a deal - and as you build the project, you kind of just figure it out.

Another thing I would maybe like to see is some like "try it in the preview" sentences to try and get users more engaged. For instance, maybe when it mentioned that password inputs are hidden it could say something like "try it in the preview to make sure the input is hidden" - or when you set the min width of something: "resize the preview to make sure it works" - perhaps we don't want any of that, I'm not sure - but I think it adds some engagement

One frustrating thing I may have mentioned on another review is that when adding something I want to see in the preview - specifically, when I was styling the submit button - I want to see it - but every time I add a property, the preview refreshes and sends me to the top of the page. So I scroll down, and add another property, and it sends me to the top again - not much we can do about that I don't think. And I suppose that's the same as the current curriculum preview.

I don't think of the above needs to be changed here - just sharing my thoughts - maybe something to keep in mind.

Like I said - this looks real good 👍 You can tell how much thought went into it.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-042.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-042.md:21]: Oops, I do not know how these got through

File: curriculum/challenges/_meta/registration-form/meta.json
Comment (ShaunSHamilton) [reply curriculum/challenges/_meta/registration-form/meta.json:5]: 🤷‍♂️

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-036.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-036.md:10]: The _Cat Photo App_ is before this. So, Campers should have an idea of what it is.

In fact, looking at the lessons in that block, this lesson is not new.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-038.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-038.md:10]: I am trying to be slightly less explicit with previously covered topics. So, I am not happy with mentioning "`text-align`".

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-040.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-040.md:19]: Firefox being different again 😠

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-046.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-046.md:10]: You do not see this

https://user-images.githubusercontent.com/51722130/134699824-28523981-6769-4dc5-b82f-55407f39e4f7.mp4

:

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-049.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-049.md:12]: I am confused... which browser are you using?

You should see this:
image

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-055.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-055.md:12]: I was going off of the mention of _"...is given to `input` elements you can type in."_ - of which, the submit input is not one.

File: curriculum/challenges/_meta/registration-form/meta.json
Comment (moT01) [reply curriculum/challenges/_meta/registration-form/meta.json:5]: I would maybe change it since 42 is the default. Well, either this one or the accessibility quiz will have to change - otherwise...

"
> The result was that the lessons jumped between the two blocks after finishing each lesson.
"Tom"

"
-Shaun

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-049.md
Comment (moT01) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-049.md:12]: Pretty sure I was using firefox. Maybe I wasn't seeing it right or something - pretty sure the color was the same before and after adding the solution - which I thought was a gray. This is quite minor - and I may have just missed something.

Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-049.md:12]: The colour is the same, but that is the point. If the Camper does not follow the instructions (e.g. by blindly copying/moving all properties) then the background colour will change.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-055.md
Comment (moT01) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-055.md:12]: My brain may have just skipped over the "you can type in" part.

I think I know what my issue here was. I was looking at the size of elements, I should have been looking at the padding. Assuming that is correct...

> you might be able to notice the file `input` is smaller than the other text `input` elements.

maybe "you might be able to notice more space around the text..." or something?

> Using another attribute selector, style the `input` with a `type` of `file` to be the same size as the other `input` elements.

"size" here made me think I should be changing the width or height, but it wants padding. So maybe "style the ... to have the same padding..."

I didn't load this back up to make sure, so I might be off - but I think that's what my problem was. The verbiage on something like this isn't a deal breaker for me to merge this. If it's a problem, the forum will let us know.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-036.md
Comment (moT01) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-036.md:10]: If you're fine with it - I'm fine with it

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-046.md
Comment (moT01) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-046.md:10]: I didn't think so (firefox) - but the change is small, I just may have missed it on the refresh. I'm fine with leaving it.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-049.md
Comment (moT01) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-049.md:12]: Sounds good - I'm fine with leaving it

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-038.md
Comment (moT01) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-038.md:10]: > I am trying to be slightly less explicit

Yea, I figured that as I went through more steps. I like that.

I think my problem here was that I thought adding the margin was supposed to be the thing that centered the text. I'm fine with leaving it, but maybe something like: "Center the `h1` and `p` element using the property and value to do that. Then, give them a `margin` of `1em auto` to add some space."

Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-038.md:10]: I think the instructions could be clearer, but I think you are confusing the concept:
- The `h1` and `p` elements are centered using `margin` - as you correctly thought
- The text can only be centered by following: _"...and center aligning the text."_

So, if you can think of a nice way to phrase that, then we can clear it up.

Comment (moT01) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-038.md:10]: I changed the suggestion.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-041.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-041.md:10]: I like the hint too much to be this explicit. As such, I will leave it as is for now.

I think it allows the Camper to think just enough.

**Comment by nhcarrigan** [comment]
image

I can't approve this 😆

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
This is coming along nicely.

I have a couple of tiny nitpicks, but the bulk of this is ready to go.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-018.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-018.md:10]: We will need to have some discussion as a team about whether to have a route that does not do this leading to an existing service. Then, we will need to adjust at least 3 of the projects to match.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-008.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-008.md:10]: Well, you are just special.

File: curriculum/challenges/english/01-responsive-web-design/registration-form/part-006.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/01-responsive-web-design/registration-form/part-006.md:33]: I was wondering about how lenient to be with these kinds of test:
- Use `trim()` to allow space at ends?
- Use `assert.match` to allow other extra stuffs?
- Use `assert.match` with `/stuff/i` to allow any casing?

**Review by moT01 — APPROVED** [review]
Still LGTM 🎉

Edit: I can approve this: `use clear, Tom-like language for 038`

**Review by naomi-lgbt — APPROVED** [review]
I'm happy with this. Nice work Shaun!

**Pull request merged**
